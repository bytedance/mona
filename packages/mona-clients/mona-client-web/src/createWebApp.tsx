import { AppConfig } from '@bytedance/mona';
import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter, Switch, Route, Redirect, useHistory } from 'react-router-dom';
import NavBar from './components/NavBar';
import TabBar from './components/TabBar';
import { formatPath, parseSearch, GLOBAL_LIFECYCLE_STORE } from '@bytedance/mona-shared';

export const WrapperComponent: React.FC<{ title: string }> = ({ children, title }) => {
  useEffect(() => {
    if (title) {
      document.title = title;
    } else {
      document.title = 'Mona Web';
    }
  }, [title]);

  return <>{children}</>;
};

const maskStyle: React.CSSProperties = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  zIndex: 1,
  backgroundColor: 'rgba(255, 255, 255)',
};

const NoMatchImgWrapperStyle: React.CSSProperties = {
  color: '#0000008C',
  fontSize: '14px',
  textAlign: 'center',
  lineHeight: '20px',
  whiteSpace: 'nowrap',
  position: 'absolute',
  left: '50%',
  top: '50%',
  transform: 'translate(-50%, -50%)',
};

const imgStyle = { width: 320, marginBottom: 8 };
export const NoMatch: React.FC<{ defaultPath: string }> = ({ defaultPath }) => {
  const history = useHistory();
  useEffect(() => {
    // app生命周期pageNotFound
    //@ts-ignore
    window[GLOBAL_LIFECYCLE_STORE]?.handlePageNotFound?.(defaultPath);
  }, [defaultPath]);
  return (
    <div style={maskStyle}>
      <div style={NoMatchImgWrapperStyle}>
        <img style={imgStyle} src={'https://lf3-fe.ecombdstatic.com/obj/ecom-open-butler/mona/error.png'} />
        <div>
          <span>
            不存在路由 {location.pathname} <a onClick={() => history.push(formatPath(defaultPath))}>返回首页</a>
          </span>
        </div>
      </div>
    </div>
  );
};

export interface PageProps {
  search: string;
  searchParams: Record<string, string>;
}

export const HistorySetWrapper: React.FC = ({ children }) => {
  const history = useHistory();

  // set global history to implement navigateTo and redirectTo api
  useEffect(() => {
    // @ts-ignore
    window.__mona_history = history;
  }, [history]);

  return <>{children}</>;
};

const defaultLightConfig: AppConfig['light'] = { mode: 'sidebar-semi-420' };

function prepareLightApp(config: AppConfig['light']) {
  // @ts-ignore
  if (typeof window.__MONA_LIGHT_APP_INIT_CB === 'function') {
    // @ts-ignore
    window.__MONA_LIGHT_APP_INIT_CB({ ...defaultLightConfig, ...config });
    // @ts-ignore
    window.__MONA_LIGHT_APP_INIT_CB = undefined;
  }
}

export function createWebApp(
  Component: React.ComponentType<any>,
  routes: { path: string; title: string; component: React.ComponentType<any> }[],
  options?: {
    tabBar?: AppConfig['tabBar'],
    navBar?: AppConfig['window'],
    defaultPath?: string
    light?: AppConfig['light']
  }
) {
  const render = ({ dom }: { dom: Element | Document }) => {
    prepareLightApp(options?.light)

    ReactDOM.render(
      <BrowserRouter>
        <HistorySetWrapper>
          <Component>
            {options?.navBar && <NavBar {...options?.navBar} />}
            <Switch>
              {routes?.map(route => (
                <Route
                  key={route.path}
                  path={formatPath(route.path)}
                  children={({ location }) => (
                    <WrapperComponent title={route.title}>
                      <route.component search={location.search} searchParams={parseSearch(location.search)} />
                    </WrapperComponent>
                  )}
                />
              ))}
              {routes?.length && (
                <Route exact path="/">
                  <Redirect to={formatPath(routes[0].path || options?.defaultPath || '/')} />
                </Route>
              )}
              <Route path="*">
                <NoMatch defaultPath={formatPath(routes[0].path || options?.defaultPath || '/')} />
              </Route>
            </Switch>
            {options?.tabBar && <TabBar tab={options?.tabBar} />}
          </Component>
        </HistorySetWrapper>
      </BrowserRouter>,
      dom.querySelector('#root'),
    );
  };

  // 在首次加载和执行时会触发该函数
  const provider = () => ({
    render, // 应用在渲染时会触发该 hook
    destroy({ dom }: { dom: Element }) {
      // 应用在销毁时会触发该 hook
      const root = (dom && dom.querySelector('#root')) || dom; // 若为 JS 入口直接将传入节点作为挂载点和销毁节点
      if (root) {
        // 做对应的销毁逻辑，保证子应用在销毁时对应的副作用也被移除
        ReactDOM.unmountComponentAtNode(root);
      }
    },
  });

  if (!window.__MARFISH__) {
    render({ dom: document });
  }

  return {
    provider,
  };
}
