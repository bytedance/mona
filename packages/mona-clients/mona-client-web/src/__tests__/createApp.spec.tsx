import { NoMatch, HistorySetWrapper, WrapperComponent } from '../createWebApp';

import React from 'react';
import { configure, mount } from 'enzyme';

import mountTest from '../../../../tests/shared/mountTest';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

describe('NoMatch', () => {
  mountTest(NoMatch);

  it('should render correctly', () => {
    expect(mount(<NoMatch defaultPath="/path/a/b" />).render()).toMatchSnapshot();
  });
});

describe('WrapperComponent', () => {
  mountTest(WrapperComponent);

  it('should render correctly', () => {
    expect(
      mount(
        <WrapperComponent title="/path/a/b">
          <div>test</div>
        </WrapperComponent>,
      ).render(),
    ).toMatchSnapshot();
  });
});

describe('HistorySetWrapper', () => {
  mountTest(HistorySetWrapper);

  it('should render correctly', () => {
    expect(
      mount(
        <HistorySetWrapper>
          <div>test</div>
        </HistorySetWrapper>,
      ).render(),
    ).toMatchSnapshot();
  });
});
