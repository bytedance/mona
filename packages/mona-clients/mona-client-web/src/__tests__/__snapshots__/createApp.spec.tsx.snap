// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistorySetWrapper should render correctly 1`] = `
LoadedCheerio {
  "0": <div>
    test
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`NoMatch should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; z-index: 1;"
  >
    <div
      style="color: rgba(0, 0, 0, 0.549); font-size: 14px; text-align: center; line-height: 20px; white-space: nowrap; position: absolute; left: 50%; top: 50%; transform: translate(-50%, -50%);"
    >
      <img
        src="https://lf3-fe.ecombdstatic.com/obj/ecom-open-butler/mona/error.png"
        style="width: 320px; margin-bottom: 8px;"
      />
      <div>
        <span>
          不存在路由 blank 
          <a>
            返回首页
          </a>
        </span>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`WrapperComponent should render correctly 1`] = `
LoadedCheerio {
  "0": <div>
    test
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
